### PR TITLE
Add option for additional assembly probe dirs

### PR DESCRIPTION
**Pull request by ViktorHofer**
Adds an optional argument to probe for additional assemblies at a different location than the test assembly's directory. This is needed for corefx but should also be useful for advanced scenarios in regular .NET apps.

cc @tonerdo

### CODE REVIEW

**Comment by tonerdo** [comment]
@ViktorHofer I believe #213 already addresses this?

**Comment by ViktorHofer** [comment]
You are absolutely right, I missed that PR. I think both PRs present individual benefits. I will incorporate my changes into @jherby2k's PR.

**Comment by ViktorHofer** [comment]
@tonerdo I force pushed @jherby2k and my changes into this PR as I couldn't update his base PR. This includes all the changes. Can you please take a look?

File: src/coverlet.core/Coverage.cs [hunk 1]
```diff
@@ -205,29 +218,29 @@ private void CalculateCoverage()
...
-                                    branchesToRemove.Add(moveNextBranch);
-                                }
-                            }
-                        }
-                    }
-                    foreach (var branchToRemove in branchesToRemove)
-                    {
-                        document.Value.Branches.Remove(branchToRemove.Key);
-                    }
+                // for MoveNext() compiler autogenerated method we need to patch false positive (IAsyncStateMachine for instance)
```
Comment (ViktorHofer) [inline src/coverlet.core/Coverage.cs]: These are white-space only changes that are wrong in upstream.

**Comment by jherby2k** [comment]
Can you just add https://github.com/ViktorHofer/coverlet/commit/3c480c72edaba748b0a299add502f89e245d8587 to this? I'm so over my head with this complex PR stuff.

**Comment by ViktorHofer** [comment]
That is already part of the changes or am I missing something?

**Comment by jherby2k** [comment]
As I said I’m completely confused at this point. I added that commit last night to my now fubar’d PR.

I’m not looking at coverlet any more at this point. if this ever makes it into a release, cool.

**Comment by tonerdo** [comment]
@jherby2k no worries, your contributions are already included in this PR. I'll do one final sweeping review, your contribution is greatly appreciated

**Comment by ViktorHofer** [comment]
@jherby2k your commits are cherry-picked into this PR, you are still the author of the commits and 50% of the work is yours. If I crossed a line please tell me, we usually work this way in our team, therefore I wasn't aware that I could be insensitive.

@tonerdo thanks a lot 👍

**Comment by jherby2k** [comment]
No no, its all good. I’m just having a frustrating day in general. Thanks for your contribution!

**Comment by ViktorHofer** [comment]
Ok got it. Hope it'll get better soon :)

File: src/coverlet.core/Helpers/InstrumentationHelper.cs [hunk 2]
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
...
+                if (!Directory.Exists(fullPath)) continue;
+
+                if (includeDirectory.EndsWith("*", StringComparison.Ordinal))
+                    dirs.AddRange(Directory.GetDirectories(fullPath));
+                else
+                    dirs.Add(fullPath);
+            }
+
+            // The test module's name must be unique.
+            var uniqueModules = new HashSet<string>
```
Comment (tonerdo) [inline src/coverlet.core/Helpers/InstrumentationHelper.cs]: This has the added benefit of preventing the test module from being included in instrumentation 👍

File: src/coverlet.core/Coverage.cs [hunk 3]
```diff
@@ -40,23 +42,33 @@ public Coverage(string module, string[] excludeFilters, string[] includeFilters,
...
+                    // Guard code path and restore if instrumentation fails.
+                    try
+                    {
+                        var result = instrumenter.Instrument();
+                        _results.Add(result);
+                    }
+                    catch (Exception)
+                    {
+                        InstrumentationHelper.RestoreOriginalModule(module, _identifier);
+                        throw;
```
Comment (tonerdo) [inline src/coverlet.core/Coverage.cs]: We shouldn't throw an exception here, instead we should let it continue with other modules. The users will simply see coverage of `0` across board. I'm looking into adding verbose logging that will communicate situations like this to the user

File: src/coverlet.core/Helpers/InstrumentationHelper.cs [hunk 4]
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
...
+            Debug.Assert(includeDirectories != null, "Parameter " + nameof(includeDirectories) + " in method " + 
+                nameof(InstrumentationHelper) + "." + nameof(GetCoverableModules) + " must not be null");
+
+            string moduleDirectory = Path.GetDirectoryName(module);
+            if (moduleDirectory == string.Empty)
+            {
+                moduleDirectory = Directory.GetCurrentDirectory();
+            }
+
+            var dirs = new List<string>(1 + includeDirectories.Length)
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: It has very subtle perf benefits but what we pass in here isn't right as some directories expand later when using the `*`. Therefore I just remove the capacity argument here.

File: src/coverlet.core/Helpers/InstrumentationHelper.cs [hunk 5]
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
...
+            {
+                // Add the test assembly's directory.
+                moduleDirectory
+            };
+
+            // Prepare all the directories in which we probe for modules.
+            foreach (var includeDirectory in includeDirectories.Where(d => d != null))
+            {
+                var fullPath = (!Path.IsPathRooted(includeDirectory)
+                    ? Path.GetFullPath(Path.Combine(moduleDirectory, includeDirectory))
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: That would have caused frustration for customers. Thanks for spotting!

**Comment by ViktorHofer** [comment]
@tonerdo Thanks for the through feedback. I addressed all your comments. Hope we can get this into master now.

Additionally I have an ask: Can we please publish a new version to either an internal feed (myget) or nuget (preferably nuget) after this is in? Thanks a lot!

File: src/coverlet.core/Helpers/InstrumentationHelper.cs [hunk 6]
```diff
@@ -40,10 +78,11 @@ public static bool HasPdb(string module)
             }
         }
 
-        public static void BackupOriginalModule(string module, string identifier)
+        public static string BackupOriginalModule(string module, string identifier)
```
Comment (tonerdo) [inline src/coverlet.core/Helpers/InstrumentationHelper.cs]: Oh! One more question here. Why did you update this to return a string as I don't see it used anywhere

**Comment by tonerdo** [comment]
Sure @ViktorHofer, will merge and publish once I get an answer to my little nit above and the CI passes

File: src/coverlet.core/Helpers/InstrumentationHelper.cs (hunk 6 above)
Comment (ViktorHofer) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: @jherby2k changed that to simplify testing: https://github.com/tonerdo/coverlet/pull/253/files#diff-009f00622e86f8b17a0f74fc89bd66a6R31

Comment (ViktorHofer) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: I'm open to revert the change if you wish so.

Comment (tonerdo) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: Please do, thanks!

Comment (ViktorHofer) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: Done. I also added the `overwrite: true` parameter as in local testing when debugging I hit that once. I hope that's fine.

Comment (tonerdo) [reply src/coverlet.core/Helpers/InstrumentationHelper.cs]: That's fine

**Comment by tonerdo** [comment]
> consider using Azure DevOps

Hehehe, shameless plug 😄

**Comment by tonerdo** [comment]
@ViktorHofer there's a new release with these changes

**Comment by ViktorHofer** [comment]
Amazing, thanks 😊

**Comment by jherby2k** [comment]
Thanks fellas!

**Pull request merged**